File: src/main/java/cal/projeteq3/glucose/model/jobOffer/JobOffer.java
package cal.projeteq3.glucose.model.jobOffer;

import cal.projeteq3.glucose.exception.unauthorizedException.CvNotApprovedException;
import cal.projeteq3.glucose.exception.unauthorizedException.JobOfferNotOpenException;
import cal.projeteq3.glucose.model.Department;
import cal.projeteq3.glucose.model.Semester;
import cal.projeteq3.glucose.model.user.Employer;
import cal.projeteq3.glucose.model.user.Student;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class JobOffer{
	@Id
	@GeneratedValue
	private Long id;

	@Column(nullable = false)
	@Enumerated(EnumType.STRING)
	private JobOfferState jobOfferState = JobOfferState.SUBMITTED;

	@Column(nullable = false)
	private String title;

	@Column(nullable = false)
	@Enumerated(EnumType.STRING)
	private Department department;

	@Column(nullable = false)
	private String location;

	@Column(nullable = false, length = 2000)
	private String description;

	private float salary;

	@Column(nullable = false)
	private LocalDate startDate;

	private int duration;

	@Column(nullable = false)
	private LocalDate expirationDate;

	private int hoursPerWeek;

	private int nbOfCandidates;

	@ToString.Exclude
	@ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
	private Employer employer;

	@OneToMany(mappedBy = "jobOffer", cascade = CascadeType.ALL, orphanRemoval = true)
	private List<JobApplication> jobApplications = new ArrayList<>();

	private String refusReason;

	@Embedded
	private Semester semester;

	public void addJobApplication(JobApplication jobApplication){
		this.jobApplications.add(jobApplication);
	}

	public void copy(JobOffer jobOffer){
		this.title = jobOffer.getTitle();
		this.department = jobOffer.getDepartment();
		this.location = jobOffer.getLocation();
		this.description = jobOffer.getDescription();
		this.salary = jobOffer.getSalary();
		this.startDate = jobOffer.getStartDate();
		this.duration = jobOffer.getDuration();
		this.expirationDate = jobOffer.getExpirationDate();
		this.jobOfferState = jobOffer.getJobOfferState();
		this.hoursPerWeek = jobOffer.getHoursPerWeek();
		this.nbOfCandidates = jobOffer.getNbOfCandidates();
		this.semester = jobOffer.getSemester();
	}

	public JobApplication apply(Student student){
		if(!student.hasApprovedCv()) throw new CvNotApprovedException();
		if(getJobOfferState() != JobOfferState.OPEN) throw new JobOfferNotOpenException();

		JobApplication jobApplication = new JobApplication();
		jobApplication.setStudent(student);
		jobApplication.setJobOffer(this);
		this.jobApplications.add(jobApplication);

		return jobApplication;
	}

	public boolean hasApplied(Long studentId){
		return this.jobApplications.stream().anyMatch(jobApplication -> jobApplication.getStudent().getId().equals(studentId));
	}

	public boolean isHiring() {
		if (this.jobOfferState.equals(JobOfferState.TAKEN)) return false;
		if (isFull()){
			this.jobOfferState = JobOfferState.TAKEN;
			return false;
		}
		if (isExpired()){
			this.jobOfferState = JobOfferState.EXPIRED;
			return false;
		}
		return isTakingIn();
	}

	private boolean isExpired(){
		return this.expirationDate.isBefore(LocalDate.now());
	}

	private boolean isFull(){
		return this.jobApplications.stream().filter(jobApplication -> jobApplication.getJobApplicationState().equals(JobApplicationState.ACCEPTED)).count() >= this.nbOfCandidates;
	}

	private boolean isTakingIn(){
		return this.jobApplications.stream().filter(jobApplication -> jobApplication.getJobApplicationState().equals(JobApplicationState.ACCEPTED)).count() < this.nbOfCandidates;
	}
}

File: frontend/src/model/JobOffer.js
class JobOffer{
	id
	title
	department
	location
	description
	salary
	startDate
	duration
	expirationDate
	jobOfferState
	hoursPerWeek
	dueDate

	constructor(){
		this.reset()
	}

	init(jobOffer){
		if(jobOffer.id) this.id = jobOffer.id
		else delete this.id
		if(jobOffer.title) this.title = jobOffer.title
		else delete this.title
		if(jobOffer.positionTitle) this.positionTitle = jobOffer.positionTitle
		else delete this.positionTitle
		if(jobOffer.department) this.department = jobOffer.department
		else delete this.department
		if(jobOffer.location) this.location = jobOffer.Location
		else delete this.location
		if(jobOffer.description) this.description = jobOffer.description
		else delete this.description
		if(jobOffer.salary) this.salary = jobOffer.salary
		else delete this.salary
		if(jobOffer.dueDate) this.dueDate = jobOffer.dueDate
		else delete this.dueDate
		if(jobOffer.expirationDate) this.expirationDate = jobOffer.expirationDate
		else delete this.expirationDate
		if(jobOffer.jobOfferState) this.jobOfferState = jobOffer.jobOfferState
		else delete this.jobOfferState
		if(jobOffer.hoursPerWeek) this.hoursPerWeek = jobOffer.hoursPerWeek
		else delete this.hoursPerWeek
    if(jobOffer.startDate) this.startDate = jobOffer.startDate
		else delete this.startDate
		if(jobOffer.duration) this.duration = jobOffer.duration
		else delete this.duration
		if(jobOffer.formation) this.formation = jobOffer.formation
		else delete this.formation
		if(jobOffer.isApproved !== undefined) this.isApproved = jobOffer.isApproved
		else this.isApproved = this.calculateApproval(jobOffer)
	}

	reset(){
		this.id = 0
		this.title = ''
		this.department = ''
		this.location = ''
		this.description = ''
		this.salary = 0
		this.startDate = ''
		this.duration = 0
		this.expirationDate = ''
		this.jobOfferState = ''
		this.hoursPerWeek = 0
		this.isApproved = false
	}

	calculateApproval(jobOffer){
		const currentDate = new Date();
		return jobOffer.jobOfferState === "OPEN" && new Date(jobOffer.expirationDate) > currentDate;
	}

}

export default JobOffer

